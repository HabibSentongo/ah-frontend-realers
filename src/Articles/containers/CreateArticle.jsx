import {
  Form, Row, Col, Button, ProgressBar,
} from 'react-bootstrap';
import React from 'react';
import '../CSS/CreateArticle.scss';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCamera } from '@fortawesome/free-solid-svg-icons';
import Loader from 'react-loader';
import PropTypes from 'prop-types';
import { storage } from '../../commons/Util/FirebaseConfig';
import postAnArticle from '../redux/actions/CreateArticleAction';
import ErrorLabel from '../../commons/components/ErrorLabel';
import SuccessLabel from '../../commons/components/SuccessLabel';
import { TITLE_REGEX, BODY_REGEX, TAGS_REGEX } from '../constants';

export class CreateArticle extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isFieldInvalid: {
        title: null,
        body: null,
        tags: null,
      },
      articleTitle: '',
      articleBody: '',
      articleTags: '',
      uploadProgress: 0,
      uploadError: 'No Error',
      displayProgress: 'none',
      isFormValid: false,
      articlePhoto: {
        name: 'No Photo selected',
        data: null,
      },
    };
  }

  componentWillMount() {
    const token = localStorage.getItem('user_token');
    if (token === null) {
      window.location = '/login';
    }

    return true;
  }

  handleCreateArticle = (event) => {
    event.preventDefault();
    const {
      articleTitle, articleBody, articleTags, articlePhoto,
    } = this.state;
    const articleData = {
      article: {
        title: articleTitle,
        body: articleBody,
        description: articleBody,
      },
    };

    if (articleTags !== '') {
      const tags = articleTags.split(',');
      articleData.article.tagName = tags;
    }

    const { postAnArticle } = this.props;

    if (articlePhoto.data !== null) {
      this.handleImageUpload();
    } else {
      postAnArticle(articleData);
    }
  }

  handleChange = (event) => {
    const { target: { name, value, files } } = event;
    const { isFieldInvalid } = this.state;

    switch (name) {
      case 'articleTitle':
        isFieldInvalid.title = !TITLE_REGEX.test(value);
        break;
      case 'articleBody':
        isFieldInvalid.body = !BODY_REGEX.test(value);
        break;
      case 'articleTags':
        isFieldInvalid.tags = !TAGS_REGEX.test(value);
        break;
      case 'articlePhoto':
        this.setState({
          articlePhoto: {
            name: 'One Photo Added, click here again to change photo',
            data: files[0],
          },
        });
        break;
      default:
        break;
    }
    if (name !== 'articlePhoto') {
      this.setState({
        [name]: value,
      });
    }

    if ((isFieldInvalid.title === false
        && isFieldInvalid.body === false
    )) {
      this.setState({
        isFormValid: true,
      });
    } else {
      this.setState({
        isFormValid: false,
      });
    }
  }

  handleImageUpload = () => {
    const {
      articleTitle, articleBody, articleTags, articlePhoto,
    } = this.state;

    const { postAnArticle } = this.props;

    this.setState({
      displayProgress: 'block',
    });

    const imageUpLoad = storage.ref(`article-image/${articlePhoto.data.name}`).put(articlePhoto.data);
    imageUpLoad.on('state_changed',
      (snapshot) => {
        const progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
        this.setState({
          uploadProgress: progress,
        });
      },
      (error) => {
        this.setState({
          uploadError: error,
        });
      },
      /* istanbul ignore next */
      (complete) => {
        storage.ref('article-image').child(articlePhoto.data.name).getDownloadURL()
          .then(url => {
            const articleData = {
              article: {
                title: articleTitle,
                body: articleBody,
                description: articleBody,
                image: url,
              },
            };
            if (articleTags !== '') {
              const tags = articleTags.split(',');
              articleData.article.tagName = tags;
            }
            postAnArticle(articleData);
          });
      });
  }

  render() {
    const {
      isFieldInvalid, articlePhoto, articleTitle, articleBody, isFormValid,
      articleTags, uploadProgress, displayProgress, uploadError,
    } = this.state;
    const {
      errorOccured,
      isPosting,
      isArticlePosted,
      createdArticle,
    } = this.props;
    if (createdArticle !== undefined) {
      setTimeout(() => {
        window.location = `/article/${createdArticle.data.slug}`;
      },
      4000);
    }
    return (
      <div className="create-article">
        <h2 className="section-title">Create Article</h2>
        <Form onSubmit={this.handleCreateArticle}>
          <Form.Group controlId="formBasicTitle">
            <Form.Control
              type="text"
              required
              name="articleTitle"
              placeholder="Article Title"
              value={articleTitle}
              onChange={this.handleChange}
              isInvalid={isFieldInvalid.title} />
            {
            isFieldInvalid.title
              ? (
                <Form.Text className="text-muted" variant="danger">
                  Article title can&apos;t be empty or contain only digits or with less than 4 characters
                </Form.Text>
              )
              : (
                <Form.Text className="text-muted" />
              )
            }
          </Form.Group>
          <Form.Group controlId="exampleForm.ControlTextarea1">
            <Form.Control
              name="articleBody"
              value={articleBody}
              onChange={this.handleChange}
              as="textarea"
              rows="15"
              placeholder="Article body"
              isInvalid={isFieldInvalid.body}
              required
            />
            {
            isFieldInvalid.body
              ? (
                <Form.Text className="text-muted">
                  Article body can&apos;t not be empty, should be atleast 100 characters long
                </Form.Text>
              )
              : (
                <Form.Text className="text-muted" />
              )
            }
          </Form.Group>
          <Form.Group controlId="formBasicTag">
            <Form.Control
              name="articleTags"
              value={articleTags}
              onChange={this.handleChange}
              type="text"
              placeholder="Article Tags"
              isInvalid={isFieldInvalid.tags} />
            {
            isFieldInvalid.tags
              ? (
                <Form.Text className="text-muted">
                  Enter tags should only be letters separated by comma like books,education
                </Form.Text>
              )
              : (
                <Form.Text className="text-muted" />
              )
            }
          </Form.Group>
          <Form.Group>
            <Form.Label htmlFor="articlePhoto" style={{ cursor: 'pointer' }}>
              <h5 className="article-photo">
                <span>
                  <FontAwesomeIcon icon={faCamera} />
                  {' '}
                  Add Photo
                </span>
                <span className="article-photo-name">{articlePhoto.name}</span>
              </h5>
              <Form.Control
                    id="articlePhoto"
                    name="articlePhoto"
                    type="file"
                    accept=".png,.jpg"
                    onChange={this.handleChange}
                    style={{ display: 'none' }}
                />
            </Form.Label>
            <Row style={{ display: displayProgress }}>
              {
                uploadError === 'No Error'
                  ? (
                    <Col sm={12}>
                      <ProgressBar now={uploadProgress} label={`${uploadProgress}%`} animated />
                    </Col>
                  )
                  : (
                    <ErrorLabel
                      message={String(uploadError)}
                      tip="try uploading the image again"
                    />
                  )
              }

            </Row>
          </Form.Group>
          <Row className="status-section">
            <Col sm={12}>
              {
                errorOccured !== undefined
                && (
                <ErrorLabel
                message={String(errorOccured.errorMessage.errors.body)}
                tip="edit your article body"
                />
                )
              }
              {
                isArticlePosted && <SuccessLabel message="Article has been successfully created" />
              }
            </Col>
          </Row>
          <Row>
            <Col sm={12} className="call-for-action">
              <Loader loaded={!isPosting} color="#00a4d3">
                <Button type="submit" disabled={!isFormValid} id="submitForm">
               Create Article
                </Button>
              </Loader>
            </Col>
          </Row>
        </Form>
      </div>
    );
  }
}

CreateArticle.protoTypes = {
  postAnArticle: PropTypes.func.isRequired,
  isArticlePosted: PropTypes.bool.isRequired,
};

export const mapStateToProps = state => ({
  isPosting: state.createArticle.isPosting,
  errorOccured: state.createArticle.error,
  isArticlePosted: state.createArticle.status,
  createdArticle: state.createArticle.item,
});

export default connect(mapStateToProps, { postAnArticle })(CreateArticle);
