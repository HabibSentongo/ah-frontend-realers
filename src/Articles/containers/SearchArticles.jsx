import React, { Component } from 'react';
import { Container } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Article from '../../commons/components/Article';
import fetchSearchedArticles from '../redux/actions/SearchArticlesAction';
import '../CSS/SearchArticles.scss';
import ErrorLabel from '../../commons/components/ErrorLabel';
import Loading from '../../commons/components/Loading';

export class SearchArticles extends Component {
  componentDidMount() {
    const { match: { params: { searchkey } }, getSearchedArticles } = this.props;

    getSearchedArticles(searchkey);
  }

  render() {
    const { items, match: { params: { searchkey } }, isSearching } = this.props;
    return (
      <div className="search-articles">
        {
        isSearching === true
          ? <Loading />
          : (
            <Container>
              {items.length !== 0 ? (
                <div>
                  <h2 className="section-title">{`Search Results For - '${searchkey}' `}</h2>
                  <div className="article-display-box">
                    {items.map(article => (
                      <Article key={article.slug} article={article} />
                    ))}
                  </div>
                </div>
              ) : (
                <ErrorLabel
                  message={`Your search - '${searchkey}' did not match any of the articles`}
                  tip="Try again with another word"
                />
              )}
            </Container>
          )
        }
      </div>
    );
  }
}

SearchArticles.propTypes = {
  items: PropTypes.array,
  match: PropTypes.object.isRequired,
  isSearching: PropTypes.bool,
  getSearchedArticles: PropTypes.func.isRequired,
};

SearchArticles.defaultProps = {
  items: [],
  isSearching: false,
};

export const mapStateToProps = state => {
  const { isSearching, items } = state.searchedArticles;
  return { isSearching, items };
};

export const mapDispatchToProps = dispatch => ({
  getSearchedArticles: (searchedKey) => {
    dispatch(fetchSearchedArticles(searchedKey));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SearchArticles);
