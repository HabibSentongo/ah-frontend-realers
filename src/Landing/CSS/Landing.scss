@import '../../commons/assets/variables/font.scss';
@import '../../commons/assets/variables/colors.scss';

.landing {
  border-top: 1px solid $light-gray;
  min-height: 250px;
}
