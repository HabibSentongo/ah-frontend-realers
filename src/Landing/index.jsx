import React, { Component } from "react";
import FacebookButton from "../auth/socialAuth/facebook/components/Facebook";
import GoogleButton from "../auth/socialAuth/google/components/Google";
import axios from "axios";
import history from '../commons/history';

export default class SocialAuthentication extends Component {
  constructor(props){
    super(props);
    this.state = {
      errorMessage: false
    };
  }
  responseFacebook = response => {
    const config = {
      headers: {
        "Content-Type": "application/json"
      },
      responseType: "json"
    };

    axios
      .post(
        "https://ah-backend-realers-staging.herokuapp.com/api/auth/facebook/",
        {
          user: {
            access_token: response.accessToken
          }
        },
        config
      )
      .then(response => {
        localStorage.setItem("access_token", response.data.user.token);
        history.push('/')
      })
      .catch(() => {
        this.setState({errorMessage: true})
      });
  };

  responseGoogle = response => {
    const config = {
      headers: {
        "Content-Type": "application/json"
      },
      responseType: "json"
    };

    axios
      .post(
        "https://ah-backend-realers-staging.herokuapp.com/api/auth/google/",
        {
          user: {
            access_token: response.Zi.id_token
          }
        },
        config
      )
      .then(response => {
        localStorage.setItem("access_token", response.data.user.token);
        history.push('/')
      })
      .catch(() => {
        this.setState({errorMessage: true})
      });
  };
  render() {
    return (
      <div className='landing'>
        {this.state.errorMessage && <p>{"Check your connection. Error Authenticating user credentials"}</p>}
        <FacebookButton responseFacebook={this.responseFacebook} />
        <p> </p>
        <GoogleButton responseGoogle={this.responseGoogle} />
      </div>
    );
  }
}
