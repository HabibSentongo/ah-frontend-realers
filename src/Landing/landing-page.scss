.landing {
    height: 300px;
    width: 40%;
    border:  #00a4d3 solid 3px;
    text-align: center;
    display: block;
    margin-left: auto;
    margin-right: auto;
    padding: 50px;
    margin-bottom: 30px;
}