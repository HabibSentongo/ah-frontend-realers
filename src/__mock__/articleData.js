/* eslint-disable import/prefer-default-export */
export const ARTICLE_DATA = [
  {
    title: 'string',
    description: 'string',
    body: 'strinfghjhhg',
    createdAt: '2019-06-19T09:16:23.749462Z',
    updatedAt: '2019-06-19T09:16:23.749489Z',
    author: {
      email: 'cedric.lusiba@andela.com',
      username: 'cdv1x',
      bio: '',
      image: '',
      gender: '',
      first_name: '',
      last_name: '',
      location: '',
      birth_date: null,
    },
    slug: 'string',
    image: 'http://string.com',
    average_rating: 0,
    tagName: [],
  },
  {
    title: 'it tgf friday',
    description: 'day of the week',
    body: "You can't miss this tdf friday in ruby",
    createdAt: '2019-06-20T13:17:46.909484Z',
    updatedAt: '2019-06-20T13:17:46.909533Z',
    author: {
      email: 'alexxsanya@gmail.com',
      username: 'alexxsanya',
      bio: '',
      image: '',
      gender: 'male',
      first_name: 'Alex',
      last_name: 'Sanya',
      location: '',
      birth_date: null,
    },
    slug: 'friday',
    image: 'https://via.placeholder.com/236x124',
    average_rating: 0,
    tagName: [],
  }];
