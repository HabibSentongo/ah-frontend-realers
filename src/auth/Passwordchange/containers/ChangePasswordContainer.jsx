import { connect } from 'react-redux';
import changePasswordAction from '../redux/actions/postAction';

const mapStateToProps = state => ({
  request_success: state.changePassword.success,
  request_errors: state.changePassword.errors,
  isLoading: state.changePassword.isLoading,
});

export default connect(mapStateToProps, { changePasswordAction });
