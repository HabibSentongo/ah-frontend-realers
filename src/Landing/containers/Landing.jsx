/* eslint-disable react/prop-types */
/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import '../CSS/Landing.scss';
import RecommendedArticles from '../components/RecommendArticles';
import LatestArticles from '../components/LatestArticles';
import fetchLatestArticles from '../redux/actions/latestArticlesAction';
import Loading from '../../commons/components/Loading';

export class Landing extends Component {
  componentWillMount() {
    this.props.fetchLatestArticles();
  }

  render() {
    const {
      latestArticles,
      recommendedArticles,
      fetchStatus,
      isFetching,
    } = this.props;

    return (
      <div className="landing">
        {
        isFetching === true
          ? <Loading />
          : (
            <div>
              <RecommendedArticles
                recommendedArticles={recommendedArticles}
                fetchStatus={fetchStatus}
              />
              <LatestArticles
                latestArticles={latestArticles}
                fetchStatus={fetchStatus}
              />
            </div>
          )
        }
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  latestArticles: state.latest_articles.items,
  recommendedArticles: state.latest_articles.items,
  fetchStatus: state.latest_articles.status,
  isFetching: state.latest_articles.is_fetching,
});

export default connect(mapStateToProps, { fetchLatestArticles })(Landing);
