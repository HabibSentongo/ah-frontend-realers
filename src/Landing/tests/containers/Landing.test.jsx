/* eslint-disable no-undef */
import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Landing, mapStateToProps } from '../../containers/Landing';
import { ARTICLE_DATA } from '../../../__mock__/articleData';

const props = {
  latestArticles: ARTICLE_DATA,
  recommendedArticles: ARTICLE_DATA,
  fetchStatus: false,
  isFetching: true,
  fetchLatestArticles: jest.fn(),
};

const fetchLatestArticlesMock = jest.fn();

describe('Landing Component', () => {
  const component = shallow(<Landing
    fetchLatestArticles={fetchLatestArticlesMock}
    latestArticles={[]}
    recommendedArticles={[]}
  />);

  it('should match the snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should match the snapshot', () => {
    const wrapper = mount(<Landing
      fetchLatestArticles={fetchLatestArticlesMock}
      latestArticles={[]}
      recommendedArticles={[]}
    />);
    wrapper.setProps(props);
    expect(wrapper).toBeDefined();
    expect(fetchLatestArticlesMock).toHaveBeenCalled();
  });

  it('should match the snapshot', () => {
    const mockedState = {
      latest_articles: {
        items: ['a'],
        status: true,
        message: 'message',
        is_fetching: true,
      },
    };
    const state = mapStateToProps(mockedState);
    expect(state.fetchStatus).toBe(true);
    expect(state.isFetching).toBe(true);
  });
});
