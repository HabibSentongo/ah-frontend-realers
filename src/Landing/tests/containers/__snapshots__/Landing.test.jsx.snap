// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing Component should match the snapshot 1`] = `
<div
  className="landing"
>
  <div>
    <RecommendedArticles
      recommendedArticles={Array []}
    />
    <LatestArticles
      latestArticles={Array []}
    />
  </div>
</div>
`;
